src/pudl/transform/eia.py

```diff
@@ -575,83 +587,59 @@ def _boiler_generator_assn(  # noqa: C901
-        the boiler-generator associations as records containing
-        plant_id_eia, generator_id, boiler_id, and unit_id_pudl
+        A dataframe containing the boiler generator associations.
 
     Raises:
-        AssertionError: If the boiler - generator association graphs are not
-            bi-partite, meaning generators only connect to boilers, and boilers
-            only connect to generators.
-        AssertionError: If all the boilers do not end up with the same unit_id
-            each year.
-        AssertionError: If all the generators do not end up with the same
-            unit_id each year.
+        AssertionError: If the boiler - generator association graphs are not bi-partite,
+            meaning generators only connect to boilers, and boilers only connect to
+            generators.
+        AssertionError: If all the boilers do not end up with the same unit_id each
+            year.
+        AssertionError: If all the generators do not end up with the same unit_id each
+            year.
     """
+    debug = context.op_config["debug"]
+    eia_settings = context.resources.dataset_settings.eia
+    eia923_years = eia_settings.eia923.years
+    eia860_years = eia_settings.eia860.years
+
     if eia923_years is None:
         eia923_years = DataSource.from_id("eia923").working_partitions["years"]
     if eia860_years is None:
         eia860_years = DataSource.from_id("eia860").working_partitions["years"]
 
-    # if eia860_years or eia923_years are still empty, we can't compile the BGA.
-    # Return the unaltered input dictionary of dataframes instead.
     if not (eia860_years and eia923_years):
-        return eia_transformed_dfs
+        logger.warning("No EIA-860 or EIA-923 years specified. Returning input BGA.")
+        return eia_transformed_dfs["clean_boiler_generator_assn_eia860"]
+
+    # Do some final data formatting and assign appropriate types:
+    eia_transformed_dfs = {
+        table_name: convert_cols_dtypes(df, data_source="eia")
+        for table_name, df in eia_transformed_dfs.items()
+    }
+
     # compile and scrub all the parts
     logger.info("Inferring complete EIA boiler-generator associations.")
-    logger.info(f"{eia_transformed_dfs.keys()=}")
-    bga_eia860 = (
-        eia_transformed_dfs["boiler_generator_assn_eia860"]
-        .copy()
-        .pipe(_restrict_years, eia923_years, eia860_years)
-        .astype(
-            {
-                "generator_id": pd.StringDtype(),
-                "boiler_id": pd.StringDtype(),
-                "plant_id_eia": int,
-            }
-        )
+    logger.debug(f"{eia_transformed_dfs.keys()=}")
+    bga_eia860 = eia_transformed_dfs["clean_boiler_generator_assn_eia860"].pipe(
+        _restrict_years, eia923_years, eia860_years
     )
     # grab the generation_eia923 table, group annually, generate a new tag
-    gen_eia923 = eia_transformed_dfs["generation_eia923"].copy()
```

Comment:
Since we don't have to worry about mutating our inputs anymore I got rid of these `copy()` calls.